pluginconfig/execute.go

```diff
@@ -78,6 +81,128 @@ func (e ExecuteOffchainConfig) Validate() error {
+// and syntactically correct (WellFormed).
+type TokenDataObserverConfig struct {
+	// Type is the type of the token data observer. You can think of different token data observers as different
+	// strategies for processing token data. For example, you can have a token data observer for USDC tokens using CCTP
+	// and different one for processing LINK token.
+	Type string `json:"type"`
+	// Version is the version of the TokenObserverConfig and the matching Observer implementation for that config.
+	// This is used to determine which version of the observer to use. Right now, we only have one version
+	// of the observer, but in the future, we might have multiple versions.
+	// This is a precautionary measure to ensure that we can upgrade the observer without breaking the existing ones.
+	// Example would be CCTPv1 using AttestationAPI and CCTPv2 using a different API or completely
+	// different strategy which requires different configuration and implementation during Observation phase.
+	// [
+	//  {
+	//    "type": "usdc-cctp",
+	//    "version": "1.0",
+	//    "attestationAPI": "http://circle.com/attestation",
+	//    "attestationAPITimeout": "1s",
+	//    "attestationAPIIntervalMilliseconds": "500ms"
+	//  },
+	//  {
+	//    "type": "usdc-cctp",
+	//    "version": "2.0",
+	//    "customCirlceAPI": "http://cirle.com/gohere",
+	//    "yetAnotherAPI": "http://cirle.com/anotherone",
+	//    "customCircleAPITimeout": "1s",
+	//    "yetAnotherAPITimeout": "500ms"
+	//  }
+	//]
+	// Having version in that JSONisn't expensive, but it could reduce the risk of breaking the observers in the future.
+	Version string `json:"version"`
+
+	*USDCCCTPObserverConfig
+}
+
+// WellFormed checks that the observer's config is syntactically correct - proper struct is initialized based on type
+func (t TokenDataObserverConfig) WellFormed() error {
+	switch t.Type {
+	case USDCCCTPHandlerType:
+		if t.USDCCCTPObserverConfig == nil {
+			return errors.New("USDCCCTPObserverConfig is empty")
+		}
+	default:
+		return errors.New("unknown token data observer type")
+	}
+	return nil
+}
+
+// Validate checks that the observer's config is semantically correct - fields are set correctly
+// depending on the config's type
+func (t TokenDataObserverConfig) Validate() error {
+	switch t.Type {
+	case USDCCCTPHandlerType:
+		if err := t.USDCCCTPObserverConfig.Validate(); err != nil {
+			return err
+		}
+	default:
+		return errors.New("unknown token data observer type " + t.Type)
+	}
+	return nil
+}
+
+const USDCCCTPHandlerType = "usdc-cctp"
+
+type USDCCCTPObserverConfig struct {
+	Tokens                 map[int]USDCCCTPTokenConfig `json:"tokens"`
+	AttestationAPI         string                      `json:"attestationAPI"`
+	AttestationAPITimeout  *commonconfig.Duration      `json:"attestationAPITimeout"`
+	AttestationAPIInterval *commonconfig.Duration      `json:"attestationAPIInterval"`
+}
+
+func (p USDCCCTPObserverConfig) Validate() error {
```

Comment:
Check these as well?
```suggestion
func (p USDCCCTPObserverConfig) Validate() error {
	if p.AttestationAPITimeout == nil {
		return errors.New("AttestationAPITimeout not set")
	}
	if p.AttestationAPIInterval == nil {
		return errors.New("AttestationAPIInterval not set")
	}
```

Reply:
I've made these to be values instead of pointers

Reply:
Does 0 need to be an allowable value? IIRC these are hardcoded in the ocr2 plugin

Reply:
The short answer, "I don't know" :D Long answer, it's gonna be part of the follow-up PR. We have hardcoded values in the v1 and I don't know how to handle defaults yet. Once I backport usdc logic it should become more obvious on how to handle that.